enclave-runtime/src/lib.rs

```diff
@@ -572,19 +570,28 @@ fn get_stf_state(
 	Ok(Stf::get_state(&mut state, trusted_getter_signed.into()))
 }
 
-fn execute_top_pool_calls<O, T>(
+/// Execute pending trusted operations for all shards until the [`max_exec_duration`] is reached.
+///
+/// For fairness, the [`max_exec_duration`] is split equally among all shards. Leftover time from
+/// the previous shard is evenly distributed to all remaining shards.
+fn exec_tops_for_all_shards<PB, SB, O, T>(
 	ocall_api: &O,
 	top_pool_getter: &T,
-	latest_onchain_header: Header,
-) -> Result<(Vec<OpaqueCall>, Vec<SignedSidechainBlock>)>
+	latest_onchain_header: &PB::Header,
+	max_exec_duration: Duration,
+) -> Result<(Vec<OpaqueCall>, Vec<SB>)>
 where
+	PB: BlockT<Hash = H256>,
+	SB: SignedBlockT<Public = sp_core::ed25519::Public, Signature = MultiSignature>,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
```

Comment:
These associated type constrictions are hideous. However, they are needed as the nested code still uses explicit types, e.g., `edd25519::unseal`. We need a proper cleanup to make our code truly generic.

Reply:
it is what it is - but now thanks to your refactoring, we can easily transform this into a more generic version 👍